### ISSUE TITLE
2FA Expiry Times

### ISSUE THREAD

**Issue by boredazfcuk** [icloud-photos-downloader/icloud_photos_downloader#229]
Hey, everyone.

Quick question... I'm having to renew my 2FA authentication token pretty much every day recently...

Is it just me, or have you guys been seeing this too?

Thanks.

**Comment by kkAtGithub** [comment]
Same problem, bro

**Comment by boredazfcuk** [comment]
I've amended my script to count how many times it runs. I've set all three of my Apple IDs to sync every hour since the iOS 14 upgrade...

I'm wondering if there's an upper limit on the amount of times a device can log-in. If Apple is forcing 2FA re-authentication after x logins. My cookie may be valid for 90 days, but it may also have a 30 login limit.

**Comment by boredazfcuk** [comment]
So, on the 16th login it asked me to re-confirm 2FA... Not great going. Will test again.

**Comment by boredazfcuk** [comment]
This time around it asked to re-confirm 2FA on the 27th login attempt.

**Comment by boredazfcuk** [comment]
2FA expired on the 27th login attempt again... Hmmm. Coincidence?

**Comment by menkej** [comment]
Very interesting... I don't run the script on a regular basis at the moment. Maybe I should setup a cronjob, too...

**Comment by boredazfcuk** [comment]
2FA needed re-confirming on 27th login attempt again. Third time in a row. There must be something limiting the number of logins per 2FA token.

I've just confirmed my latest 2FA token via SMS, rather than `two-factor authentication code` to see if this increases longevity. Not holding out hope.

**Comment by boredazfcuk** [comment]
Failed on 27th again. I think I may change the defaults on my container to sync every 24hrs, or go back to getting the user to specify the interval hour, and have crontab run once a day.

**Comment by menkej** [comment]
> Failed on 27th again. I think I may change the defaults on my container to sync every 24hrs, or go back to getting the user to specify the interval hour, and have crontab run once a day.

Why do you sync so often anyways? I just want to understand your use case. IMHO the main purpose of this script is "backup" and a daily run (or even less frequent) is about "normal". So in that case it would be about once a month, which is ok for me. If we really need to support very frequent runs and the API does not support it, I'm currently without idea how to fix it... suggestions, anyone?

**Comment by boredazfcuk** [comment]
There was an issue where I would see `private db access disabled for this account` if I synced too many times in quick succession. Initially I suspected Apple was throttling connections in some way, but I think it turned out to be related to multi-threaded downloads: https://github.com/icloud-photos-downloader/icloud_photos_downloader/issues/163 and was fixed when it was removed.

So I dropped it to every hour, just because I could... That way it would be more like a synchronous copy than a backup.

It only became an issue around the time I upgraded to iOS14, but that may be coincidental. I noticed that I'd get the 2FA prompt every day... Then someone logged an issue against my container who is running it every 15mins!

I thought I'd just mention it here, to make you aware of this behaviour I'd seen, just in case it's something to do with the script operation... like a cache or temporary file that may be causing a problem.

**Comment by menkej** [comment]
> [...] I noticed that I'd get the 2FA prompt every day... Then someone logged an issue against my container who is running it every 15mins!

Thanks for sharing this! Seriously I did not expect people to run it ever 15 mins! Apple should finally publish an iCloud client for Linux... ;-) We should keep this issue open and add related info if we find it. Maybe it would be something to add to the FAQ for now that it has side effects when you run the script very often (throttling, 2FA, ...)

**Comment by w1zz4** [comment]
> > Failed on 27th again. I think I may change the defaults on my container to sync every 24hrs, or go back to getting the user to specify the interval hour, and have crontab run once a day.
> 
> Why do you sync so often anyways? I just want to understand your use case. IMHO the main purpose of this script is "backup" and a daily run (or even less frequent) is about "normal". So in that case it would be about once a month, which is ok for me. If we really need to support very frequent runs and the API does not support it, I'm currently without idea how to fix it... suggestions, anyone?

Well people like me want to merge their photos from multiple iCloud accounts and upload them to a third party cloud provider so me and my wife have access to each other photos.

It is not possible to use app-specific password to access icloud photos I bet?

**Comment by AndreyNikiforov** [comment]
> It is not possible to use app-specific password to access icloud photos I bet?

I haven't used app-specific passwords yet. From the doc on the apple site it sounds as if they are supposed to work without any modifications to the apps (kinda multiple passwords for one account). If you have app-specific passwords setup, can you please try and capture details if they do not work?

**Comment by menkej** [comment]
> > It is not possible to use app-specific password to access icloud photos I bet?
> 
> [...] If you have app-specific passwords setup, can you please try and capture details if they do not work?

I just tried. Logging in using App PW leads to an "Unknown reason" error in the base lib we're using. It does not seem to work at the moment...

```
2021-01-09 10:12:51 DEBUG    Authenticating...
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pyicloud_ipd/base.py", line 209, in authenticate
    req = self.session.post(
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 578, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/pyicloud_ipd/base.py", line 100, in request
    self._raise_error(code, reason)
  File "/usr/local/lib/python3.9/site-packages/pyicloud_ipd/base.py", line 122, in _raise_error
    raise api_error
pyicloud_ipd.exceptions.PyiCloudAPIResponseError: Unknown reason

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/joerg/Desktop/Entwicklung/python/icloud_photos_downloader/./icloudpd.py", line 5, in <module>
    main()
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/joerg/Desktop/Entwicklung/python/icloud_photos_downloader/icloudpd/base.py", line 255, in main
    icloud = authenticate(
  File "/Users/joerg/Desktop/Entwicklung/python/icloud_photos_downloader/icloudpd/authentication.py", line 29, in authenticate
    icloud = pyicloud_ipd.PyiCloudService(
  File "/usr/local/lib/python3.9/site-packages/pyicloud_ipd/base.py", line 193, in __init__
    self.authenticate()
  File "/usr/local/lib/python3.9/site-packages/pyicloud_ipd/base.py", line 216, in authenticate
    raise PyiCloudFailedLoginException(msg, error)
pyicloud_ipd.exceptions.PyiCloudFailedLoginException: ('Invalid email/password combination.', PyiCloudAPIResponseError('Unknown reason'))
```

Maybe I try again using the most current version of pyicloud later. But this it not working 100% yet. I'm still in the process of porting back from our fork...

**Comment by harrisoff** [comment]
If I run cron_script.sh manually it works fine, if cron job runs it, 2fa is required everytime

**Comment by menkej** [comment]
> If I run cron_script.sh manually it works fine, if cron job runs it, 2fa is required everytime

Is it possible that you're running the cronjob under a different user? What is `cron_script.sh` ?

**Comment by harrisoff** [comment]
> > If I run cron_script.sh manually it works fine, if cron job runs it, 2fa is required everytime
> 
> Is it possible that you're running the cronjob under a different user? What is `cron_script.sh` ?

`cron_script.sh` is from the template [cron_script.sh.example](https://github.com/icloud-photos-downloader/icloud_photos_downloader/blob/master/cron_script.sh.example).

I just found that manually running is also able to trigger two step auth sometimes. I started to use this tool yesterday, but I have already went through 2fa several times.

The document says the interval is "two months", I don't know whether this is from official Apple documents. I guess the expiry time varys for different users, depending on Apple's strategy.

**Comment by w1zz4** [comment]
> The document says the interval is "two months", I don't know whether this is from official Apple documents. I guess the expiry time varys for different users, depending on Apple's strategy.

I am pretty sure this is obsolete information and I don't think it is user dependent. It's probably still 2 months or so but also there seems to have a maximum amount of login before it expire.

**Comment by boredazfcuk** [comment]
> The document says the interval is "two months", I don't know whether this is from official Apple documents. I guess the expiry time varys for different users, depending on Apple's strategy.

You can check for yourself. Just open your cookie file and look for the `expires=` dates.

There will be one or two cookies in there. The cookie that starts with:
```
Set-Cookie3: X_APPLE_WEB_KB-
```
This is the icloud.com cookie's expiry time.

If you use two factor authentication then you will have a second cookie on a line that starts with:
```
Set-Cookie3: X-APPLE-WEBAUTH-HSA-TRUST-
```
This is the 2FA cookie and when it reaches that expiry time, you'll need to re-do the 2FA authentication.

**Comment by boredazfcuk** [comment]
>  there seems to have a maximum amount of login before it expire.

I was seeing this last month. However, I've noticed I'm up to 60+ logins on one of my containers now. Not sure the number of logins is a factor any more.

**Comment by harrisoff** [comment]
> > The document says the interval is "two months", I don't know whether this is from official Apple documents. I guess the expiry time varys for different users, depending on Apple's strategy.
> 
> You can check for yourself. Just open your cookie file and look for the `expires=` dates.
> 
> There will be one or two cookies in there. The cookie that starts with:
> 
> ```
> Set-Cookie3: X_APPLE_WEB_KB-
> ```
> 
> This is the icloud.com cookie's expiry time.
> 
> If you use two factor authentication then you will have a second cookie on a line that starts with:
> 
> ```
> Set-Cookie3: X-APPLE-WEBAUTH-HSA-TRUST-
> ```
> 
> This is the 2FA cookie and when it reaches that expiry time, you'll need to re-do the 2FA authentication.

Yes there are two cookies, icloud cookie's expires is `2021-03-15 16:18:09Z` and 2FA cookie's expires is `2021-04-14 10:48:30Z`.

However after I checked the cookies, the cron job ran and my iPhone still got a 2FA authentication.

I generated those cookies several hours ago and only used them maybe one or two times.

**Comment by boredazfcuk** [comment]
Is there anything in the syslog around the time the cron job launches?

**Comment by ronny332** [comment]
Things happen in a fully unexpected way, seems like Apple want's to fool the users a bit, who don't use the official software.

In my case, I have to synchronize 3 accounts, all with 2FA enabled. One account with 18.000 pictures, one with 3.000 and the last one with about 400 images.
The account with the lowest amount of pictures receives the most 2FA renewal requests, but only with gaps of more than 30 minutes between each request. Running requests every 5th minutes worked after hours, but this delay is much too short in my eyes and after about 2-4 hours, an 2FA renewal gets fired.
For the other accounts, it's completely different. Running a request once an hour (currently I use 6 hours) works very well, running them faster results in 2FA renewal requests.

I have to monitor the behavior for a longer time, but it doesn't seem to follow an always identical scheme.

**Comment by boredazfcuk** [comment]
I just set all my accounts (4) to every 12 hours, pretty sure once a week would probably be fine.

All my photos are uploaded to iCloud immediately, so it doesn't matter if the kids lose their phones or break the iPad. The photos should already be safe and there's no rush to download them.

I'm actually going to close this off now, as I wee seeing the 2FA cookie expire every day at one point. I don't sync so often any more and now it all works fine for me.

**Comment by petebocken** [comment]
I experienced the same issue. Last night I setup my script to automatically run every hour (Windows Task Scheduler) for my wife's account. She told me some time today she started getting auth codes pop up every so often (presumably every hour since I was running my script automatically). I checked and indeed it needed the auth code to run again just after maybe 12 or so runs. I changed my script to only run every 24 hours. I'll see how that goes.
